File: book.exchange/src/main/java/com/dominykas/book/exchange/service/NoticeService.java
package com.dominykas.book.exchange.service;

import com.dominykas.book.exchange.dto.noticeDTO.PageResponseDTO;
import com.dominykas.book.exchange.dto.noticeDTO.NoticeFilterDTO;
import com.dominykas.book.exchange.dto.noticeDTO.NoticeRequestDTO;
import com.dominykas.book.exchange.dto.noticeDTO.NoticeResponseDTO;
import com.dominykas.book.exchange.entity.Notice;
import com.dominykas.book.exchange.entity.Publication;
import com.dominykas.book.exchange.entity.User;
import com.dominykas.book.exchange.entity.enums.NoticeStatus;
import com.dominykas.book.exchange.mapper.NoticeMapper;
import com.dominykas.book.exchange.repository.NoticeRepository;
import com.dominykas.book.exchange.repository.PublicationRepository;
import com.dominykas.book.exchange.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class NoticeService {

    private final NoticeRepository noticeRepository;
    private final UserRepository userRepository;
    private final PublicationRepository publicationRepository;

    public NoticeResponseDTO createNotice(NoticeRequestDTO dto) {
        Notice notice = NoticeMapper.fromDto(dto);

        User poster = userRepository.findById(dto.getPosterId())
                .orElseThrow(() -> new RuntimeException("User not found: " + dto.getPosterId()));

        Publication publication = publicationRepository.findById(dto.getPublicationId())
                .orElseThrow(() -> new RuntimeException("Publication not found: " + dto.getPublicationId()));

        notice.setPoster(poster);
        notice.setPublication(publication);
        notice.setTimePosted(LocalDate.now());

        if (notice.getStatus() == null) {
            notice.setStatus(NoticeStatus.OPEN);
        }

        Notice saved = noticeRepository.save(notice);
        return NoticeMapper.toDto(saved);
    }

    public PageResponseDTO<NoticeResponseDTO> getNotices(
            int page,
            int size,
            String sortBy,
            String sortDir,
            NoticeFilterDTO filter
    ) {
        String mappedSortBy = mapSortBy(sortBy);

        Sort secondarySort = sortDir.equalsIgnoreCase("asc")
                ? Sort.by(mappedSortBy).ascending()
                : Sort.by(mappedSortBy).descending();

        Sort sort = Sort.by(Sort.Order.desc("status")).and(secondarySort);

        Pageable pageable = PageRequest.of(page, size, sort);

        Specification<Notice> specification = Specification
                .where(NoticeSpecification.hasPosterId(filter.getPosterId()))
                .and(NoticeSpecification.titleContains(filter.getTitle()))
                .and(NoticeSpecification.authorContains(filter.getAuthor()))
                .and(NoticeSpecification.hasLanguage(filter.getLanguage()))
                .and(NoticeSpecification.releaseYearFrom(filter.getReleaseYearFrom()))
                .and(NoticeSpecification.releaseYearTo(filter.getReleaseYearTo()))
                .and(NoticeSpecification.minPageCount(filter.getMinPageCount()))
                .and(NoticeSpecification.maxPageCount(filter.getMaxPageCount()))
                .and(NoticeSpecification.hasCover(filter.getCover()))
                .and(NoticeSpecification.hasColored(filter.getColored()))
                .and(NoticeSpecification.postedFrom(filter.getPostedFrom()))
                .and(NoticeSpecification.postedTo(filter.getPostedTo()));

        Page<Notice> noticePage = noticeRepository.findAll(specification, pageable);

        List<NoticeResponseDTO> content = noticePage.getContent()
                .stream()
                .map(NoticeMapper::toDto)
                .toList();

        return new PageResponseDTO<>(
                content,
                noticePage.getNumber(),
                noticePage.getSize(),
                noticePage.getTotalElements(),
                noticePage.getTotalPages(),
                noticePage.isFirst(),
                noticePage.isLast()
        );
    }

    public NoticeResponseDTO getNoticeById(Long id) {
        Notice notice = noticeRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Notice not found: " + id));

        return NoticeMapper.toDto(notice);
    }

    public NoticeResponseDTO getNoticeByPublicationId(Long publicationId) {
        Notice notice = noticeRepository.findByPublicationId(publicationId)
                .orElseThrow(() -> new RuntimeException("Notice not found for publication id: " + publicationId));

        return NoticeMapper.toDto(notice);
    }

    public List<NoticeResponseDTO> getNoticesByPosterId(Long posterId) {
        return noticeRepository.findAllByPosterId(posterId)
                .stream()
                .map(NoticeMapper::toDto)
                .toList();
    }

    private String mapSortBy(String sortBy) {
        return switch (sortBy) {
            case "id" -> "id";
            case "timePosted" -> "timePosted";
            default -> "timePosted";
        };
    }
}

File: frontend/app/src/app/core/services/notice.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  NoticeFilter,
  NoticeRequestDto,
  NoticeResponseDto,
  PageResponseDto
} from '../models/notice.model';

@Injectable({
  providedIn: 'root'
})
export class NoticeService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = 'http://localhost:8080/notices';

  createNotice(dto: NoticeRequestDto): Observable<NoticeResponseDto> {
    return this.http.post<NoticeResponseDto>(this.baseUrl, dto);
  }

  getNotices(
    page = 0,
    size = 12,
    sortBy = 'timePosted',
    sortDir = 'desc',
    filters?: NoticeFilter
  ): Observable<PageResponseDto<NoticeResponseDto>> {
    let params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sortBy', sortBy)
      .set('sortDir', sortDir);

    if (filters?.posterId != null) {
      params = params.set('posterId', filters.posterId);
    }

    if (filters?.title) {
      params = params.set('title', filters.title);
    }

    if (filters?.author) {
      params = params.set('author', filters.author);
    }

    if (filters?.language) {
      params = params.set('language', filters.language);
    }

    if (filters?.releaseYearFrom) {
      params = params.set('releaseYearFrom', filters.releaseYearFrom);
    }

    if (filters?.releaseYearTo) {
      params = params.set('releaseYearTo', filters.releaseYearTo);
    }

    if (filters?.minPageCount != null) {
      params = params.set('minPageCount', filters.minPageCount);
    }

    if (filters?.maxPageCount != null) {
      params = params.set('maxPageCount', filters.maxPageCount);
    }

    if (filters?.cover) {
      params = params.set('cover', filters.cover);
    }

    if (filters?.colored != null) {
      params = params.set('colored', filters.colored);
    }

    if (filters?.postedFrom) {
      params = params.set('postedFrom', filters.postedFrom);
    }

    if (filters?.postedTo) {
      params = params.set('postedTo', filters.postedTo);
    }

    return this.http.get<PageResponseDto<NoticeResponseDto>>(this.baseUrl, { params });
  }

  getNoticeById(id: number): Observable<NoticeResponseDto> {
    return this.http.get<NoticeResponseDto>(`${this.baseUrl}/${id}`);
  }

  getNoticeByPublicationId(publicationId: number): Observable<NoticeResponseDto> {
    return this.http.get<NoticeResponseDto>(`${this.baseUrl}/publication/${publicationId}`);
  }

  getNoticesByPosterId(posterId: number): Observable<NoticeResponseDto[]> {
    return this.http.get<NoticeResponseDto[]>(`${this.baseUrl}/poster/${posterId}`);
  }
}
